plugins/test/dynamic_test.cc

```diff
@@ -118,6 +120,132 @@ class LogTestBounds {
  private:
   ContextOptions& options_;
 };
+
+// Class that manages the state for a background stream running in a benchmark.
+class BackgroundStream {
+ public:
+  enum class NextAction {
+    kCreate,
+    kSendRequestHeaders,
+    kSendRequestBody,
+    kSendResponseHeaders,
+    kSendResponseBody,
+    kDestroy,
+  };
+
+  BackgroundStream(std::shared_ptr<proxy_wasm::PluginHandleBase> handle,
+                   const TestHttpContext::Headers* request_headers,
+                   const TestHttpContext::Headers* response_headers,
+                   const std::vector<std::string>* request_body_chunks,
+                   const std::vector<std::string>* response_body_chunks)
+      : handle_(handle),
+        request_headers_(request_headers),
+        response_headers_(response_headers),
+        request_body_chunks_(request_body_chunks),
+        response_body_chunks_(response_body_chunks) {}
+  BackgroundStream(BackgroundStream&& other) = default;
+  ~BackgroundStream() = default;
+
+  // Advance the state of the background stream, e.g. by writing headers or a
+  // body chunk.
+  absl::Status Advance();
+
+  NextAction next_action() { return next_action_; }
+
+ private:
+  std::shared_ptr<proxy_wasm::PluginHandleBase> handle_;
+  const TestHttpContext::Headers* request_headers_;
+  const TestHttpContext::Headers* response_headers_;
+  const std::vector<std::string>* request_body_chunks_;
+  const std::vector<std::string>* response_body_chunks_;
+  std::deque<std::string> remaining_request_body_chunks_;
+  std::deque<std::string> remaining_response_body_chunks_;
+  std::unique_ptr<TestHttpContext> stream_;
+  NextAction next_action_ = NextAction::kCreate;
+};
+
+absl::Status BackgroundStream::Advance() {
+  while (true) {  // cycle through actions until one is performed
+    switch (next_action_) {
+      case NextAction::kCreate:
+        stream_ = std::make_unique<TestHttpContext>(handle_);
+        if (request_body_chunks_ != nullptr) {
+          remaining_request_body_chunks_.assign(request_body_chunks_->begin(),
+                                                request_body_chunks_->end());
```

Comment:
I don't remember the ownership model of body chunks, and I'm a little worried that this class refers to chunks that may be later consumed by the "main streams". Is that a real issue?

Either way please add a comment somewhere in this class explaining the ownership model.

Reply:
> I don't remember the ownership model of body chunks, and I'm a little worried that this class refers to chunks that may be later consumed by the "main streams". Is that a real issue?

It shouldn't be an issue--DynamicTest::BenchHttpHandlers() creates copies of body data with each loop iteration ([code](https://github.com/GoogleCloudPlatform/service-extensions/blob/09182c7881b41ae4769be36547e93b52531949ee/plugins/test/dynamic_test.cc#L394-L395)), while BackgroundStream holds a pointer to the original body data vector that is only used as the source for copies.

Added a comment to the fields about ownership.